farm-boosted-rewards-subscriber/src/buy_mex.rs

```diff
@@ -0,0 +1,195 @@
+use auto_farm::common::unique_payments::UniquePayments;
+use subscription_fee::subtract_payments::Epoch;
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+pub type Percentage = u32;
+pub const TOTAL_PERCENTAGE: Percentage = 10_000;
+
+#[derive(TypeAbi, TopEncode, TopDecode)]
+pub struct MexActionsPercentages {
+    pub lock: Percentage,
+    pub fees: Percentage,
+    pub mex_burn: Percentage,
+}
+
+pub struct MexActionsValue<M: ManagedTypeApi> {
+    pub lock: BigUint<M>,
+    pub fees: BigUint<M>,
+    pub mex_burn: BigUint<M>,
+}
+
+impl MexActionsPercentages {
+    pub fn is_valid(&self) -> bool {
+        self.lock + self.fees + self.mex_burn == TOTAL_PERCENTAGE
+    }
+
+    pub fn get_amounts_per_category<M: ManagedTypeApi>(
+        &self,
+        total: &BigUint<M>,
+    ) -> MexActionsValue<M> {
+        let lock_amount = total * self.lock / TOTAL_PERCENTAGE;
+        let fees_amount = total * self.fees / TOTAL_PERCENTAGE;
+        let mex_burn_amount = total - &lock_amount - &fees_amount;
+
+        MexActionsValue {
+            lock: lock_amount,
+            fees: fees_amount,
+            mex_burn: mex_burn_amount,
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> MexActionsValue<M> {
+    pub fn get_total_mex(&self) -> BigUint<M> {
+        &self.lock + &self.mex_burn
+    }
+}
+
+#[multiversx_sc::module]
+pub trait BuyMexModule {
+    #[only_owner]
+    #[endpoint(addMexPair)]
+    fn add_mex_pair(&self, token_id: TokenIdentifier, pair_address: ManagedAddress) {
+        require!(token_id.is_valid_esdt_identifier(), "Invalid token ID");
+        require!(
+            self.blockchain().is_smart_contract(&pair_address),
+            "Invalid pair address"
+        );
+
+        self.mex_pairs(&token_id).set(pair_address);
+    }
+
+    #[only_owner]
+    #[endpoint(removeMexPair)]
+    fn remove_mex_pair(&self, token_id: TokenIdentifier) {
+        self.mex_pairs(&token_id).clear();
+    }
+
+    fn perform_mex_operations(
+        &self,
+        user_address: ManagedAddress,
+        token_id: TokenIdentifier,
+        total_tokens: BigUint,
+        actions_percentages: &MexActionsPercentages,
+    ) {
+        let actions_value = actions_percentages.get_amounts_per_category(&total_tokens);
+        let total_mex_to_buy = actions_value.get_total_mex();
+
+        if actions_value.fees > 0 {
+            self.total_fees().update(|fees| {
+                fees.add_payment(EsdtTokenPayment::new(
+                    token_id.clone(),
+                    0,
+                    actions_value.fees.clone(),
+                ))
+            });
+        }
+
+        let opt_bought_mex = self.buy_mex(token_id.clone(), total_mex_to_buy.clone());
+        if opt_bought_mex.is_none() {
+            // TODO: Maybe do something else with it? Discuss.
+            self.total_fees().update(|fees| {
```

Comment:
why do we have fees here and on line 81 as well ?

Reply:
Fee has to be taken only once. What is this Option for bought mex ? As I see further down the line, you buy and burn what is needed.

Reply:
In case there is no pair set for the given token.

Reply:
Why did you make `buy_mex` an `Option`? 
Why not crash the execution if `pair_mapper.is_empty()` or return an empty payment if `amount == 0` `EsdtTokenPayment` or even do nothing?